### ISSUE TITLE
DB Connection / release error

### ISSUE THREAD

**Issue by CameronSima** [seapagan/fastapi-template#331]
**Describe the bug**

**To Reproduce**
multiple requests which hit the db simultaneously or in quick succession

Seeing error:

       |     user_data = await db.fetch_one(
       |   File "/usr/local/lib/python3.9/site-packages/databases/core.py", line 152, in fetch_one
       |     return await connection.fetch_one(query, values)
       |   File "/usr/local/lib/python3.9/site-packages/databases/core.py", line 251, in __aexit__
       |     await self._connection.release()
      |   File "/usr/local/lib/python3.9/site-packages/databases/backends/postgres.py", line 174, in release
     |     assert self._database._pool is not None, "DatabaseBackend is not running"
      | AssertionError: DatabaseBackend is not running

**Comment by CameronSima** [comment]
I think there is an error in the way connections are handled in this repo. I don' think the connection should be created and disconnected with each request as is done now. I think instead of 

```
async def get_database():
    """Return the database connection as a Generator."""
    database.connect()
    yield database
    database.disconnect()
```

We should simply be yielding the connection:
```
async def get_database():
    """Return the database connection as a Generator."""
    yield database
```

And instead properly initialize and tear down the connection pool at startup/ shutdown:

```
@app.on_event("startup")
async def startup():
    await database.connect()

@app.on_event("shutdown")
async def shutdown():
    await database.disconnect()
```

**Comment by seapagan** [comment]
@CameronSima Yes you are correct, and that was the way it was done originally in the repo, I honestly can't remember why I changed it! Good catch though, thank you.

However, I have deprecated using `encode/databases` and converted to `SQLAlchemy 2` - this work is in the `devel` branch and will probably be merged to `main` once I finish the tests. I'm not at my main development PC right now but I'll check later

I'll look at any effects caused by changing the `main` branch back to fix this, I intend to freeze a release just before merging the new code so it is available for those who prefer to stay with `encode/databases`

Incidentally, the `@app.on_event` handlers are deprecated in FastAPI, the new recommended method is to use the [Lifespan Events](https://fastapi.tiangolo.com/advanced/events/)

**Comment by CameronSima** [comment]
Gotcha, I did see the lifespan events but it wasn't immediately clear to me how to do the teardown/ shutdown hook. Look forward to the new version -- nice repo/ template btw! I'm using it for a mobile app backend I'm developing.

Thanks,
Cam

**Comment by seapagan** [comment]
@CameronSima re the lifespan, setup and teardown are done in the same function, with the `setup` phase first, then a simple `yield` command, and then the `teardown` commands after that. It's a different mindset but I think I prefer it.

If you have the time, can you please try out the `devel` branch or the latest alpha tag since things have changed quite a bit with database access? TIA :)

Thanks for the kind words :blush:

**Comment by seapagan** [comment]
@CameronSima, this has been changed in the `0.4.2` branch back to how you suggested; please let me know if that works. I have merged the dev work into main, so the database and more have changed, you are probably best sticking to the 0.4.2 branch unless the migration of your own database code is posssible.
